tools/block-generator/generator/config.go

```
@@ -0,0 +1,323 @@
+	AssetCreateFraction  float32 `yaml:"asset_create_fraction"`
+	AssetDestroyFraction float32 `yaml:"asset_destroy_fraction"`
+	AssetOptinFraction   float32 `yaml:"asset_optin_fraction"`
+	AssetXferFraction    float32 `yaml:"asset_xfer_fraction"`
+	AssetCloseFraction   float32 `yaml:"asset_close_fraction"`
+
+	// App kind TX Distribution
+	AppSwapFraction  float32 `yaml:"app_swap_fraction"`
+	AppBoxesFraction float32 `yaml:"app_boxes_fraction"`
+
+	// App Swap TX Distribution
+	AppSwapCreateFraction float32 `yaml:"app_swap_create_fraction"`
+	AppSwapUpdateFraction float32 `yaml:"app_swap_update_fraction"`
+	AppSwapDeleteFraction float32 `yaml:"app_swap_delete_fraction"`
+	AppSwapOptinFraction  float32 `yaml:"app_swap_optin_fraction"`
+	AppSwapCallFraction   float32 `yaml:"app_swap_call_fraction"`
+	AppSwapCloseFraction  float32 `yaml:"app_swap_close_fraction"`
+	AppSwapClearFraction  float32 `yaml:"app_swap_clear_fraction"`
+
+	// App Boxes TX Distribution
+	AppBoxesCreateFraction float32 `yaml:"app_boxes_create_fraction"`
+	AppBoxesUpdateFraction float32 `yaml:"app_boxes_update_fraction"`
+	AppBoxesDeleteFraction float32 `yaml:"app_boxes_delete_fraction"`
+	AppBoxesOptinFraction  float32 `yaml:"app_boxes_optin_fraction"`
+	AppBoxesCallFraction   float32 `yaml:"app_boxes_call_fraction"`
+	AppBoxesCloseFraction  float32 `yaml:"app_boxes_close_fraction"`
+	AppBoxesClearFraction  float32 `yaml:"app_boxes_clear_fraction"`
+}
+
+// ---- construction and validation ----
+
+// initializeConfigFile reads the config file and validates its parameters. Certain missing
+// parameters are defaulted to a reasonable value when missing, or when an entire associated
+// group is missing.
+func initializeConfigFile(configFile string) (config GenerationConfig, err error) {
+	var data []byte
+	data, err = os.ReadFile(configFile)
+	if err != nil {
+		return
+	}
+	err = yaml.Unmarshal(data, &config)
+	if err != nil {
+		return
+	}
+
+	err = config.validateWithDefaults(true)
+	return
+}
```

moved from `generate.go` and using `validateWithDefaults()` to fail fast.